core/server/master/src/main/java/alluxio/master/metastore/rocks/RocksStore.java

```diff
@@ -94,20 +96,30 @@ public RocksStore(String name, String dbPath, String checkpointPath,
    * @return the underlying rocksdb instance. The instance changes when clear() is called, so if the
    *         caller caches the returned db, they must reset it after calling clear()
    */
-  public synchronized RocksDB getDb() {
-    return mDb;
+  public RocksDB getDb() {
+    long stamp = mStampedLock.readLock();
```

Comment:
You probably want to use an optimistic lock here, and fall back to the normal read lock in case of failure, otherwise the contention on locking and unlocking will be the same as before.

I am not sure this is better than an atomic reference either since the problem of reading from a closed db still exists (i.e. this follows the same locking logic as before) and I wonder if it is better to just let the client fail fast and fall over to the new master, instead of waiting for the lock?

Reply:
New commit have been switched to optimistic lock and pessimistic read lock

It is different from the old synchronized implementation. In the scenario of multithreading, such as InodeSyncStream sync triggered by listStatus(ExecutorService mMetadataSyncService from InodeSyncStream  will have many threads on large server), there are many threads to read and write rocksdb, and synchronized will consume a lot of time due to multithreading competition, and the read-write lock is much better. 

AtomicRef with Fail fast, in a multithreading scenario, may cause a large number of exception logs, which is inconvenient to locate the real problem. (For example, Hive is very bad at doing this)

Reply:
Another reason why AtomicRef cannot be used is that rocksdb cannot be modified during checkpoint. 

Alluxio provides two guarantees, one is that all rpc services are locked by a global lock, and the other is that rocksdb db instance cannot be got during chekcpoint because it is synchronized. 

If it is changed to AtomicRef, then background tasks such as PersistenceChecker in DefaultFileSystemMaster can modify rocksdb during checkpoint, which is a problem